ar71xx: add support for Fritz!WLAN Repeater 1750E

blocktrron opened the pull request:
This adds support for the AVM Fritz!WLAN Repeater 1750E

Tested and working:
 - 2.4GHz WiFi (correct MAC)
 - Ethernet
 - Installation via EVA bootloader
 - OpenWRT sysupgrade
 - Buttons
 - LEDs

Not working:
 - 5GHz Radio not detected on the PCIe Bus (Not planned due to ath10k not working reliably with 64MB RAM)

The 5GHz Radio seems to have a rather strange reset procedure which i weren't able to imitate in OpenWRT. Radio and PCI Bus are reset using GPIOs.

Open-source code from the manufacturer can be found [here](https://osp.avm.de/fritzwlan/fritzwlan-repeater-1750e/)

blocktrron committed:
ar71xx: add support for Fritz!WLAN Repeater 1750E

This commit adds support for the AVM Fritz!WLAN Repeater 1750E

SOC:	Qualcomm QCA9556 (Scorpion) 720MHz MIPS74Kc
RAM:    64MB Zentel A3R12E40CBF DDR2
FLASH:  16MiB Winbond W25Q128 SPI NOR
WLAN1:  QCA9556 2.4 GHz 802.11b/g/n 3x3
WLAN2:  QCA9880 5 GHz 802.11 n/ac 3x3
INPUT:  WPS button
LED:    Power, WiFi, LAN, RSSI indicator
Serial: Header Next to Black metal shield
        Pinout is 3.3V - RX - TX - GND (Square Pad is 3.3V)
        The Serial setting is 115200-8-N-1.

Tested and working:
 - Ethernet
 - 2.4 GHz WiFi (correct MAC)
 - 5 GHz WiFi (correct MAC)
 - Installation via EVA bootloader
 - OpenWRT sysupgrade
 - Buttons
 - LEDs

Installation via EVA:
In the first seconds after Power is connected, the bootloader will
listen for FTP connections on 192.168.178.1. Firmware can be uploaded
like following:

  ftp> quote USER adam2
  ftp> quote PASS adam2
  ftp> binary
  ftp> debug
  ftp> passive
  ftp> quote MEDIA FLSH
  ftp> put openwrt-sysupgrade.bin mtd1

Note that this procedure might take up to two minutes.
You need to powercycle the Device afterwards to boot OpenWRT.

Signed-off-by: David Bauer <mail@david-bauer.net>

robimarko commented:
What are the specifications of the device?
Primarily AC card in it and eth PHY

ath10k will work on 64Mb,there are lots of devices with that combo

blocktrron commented:
@robimarko Full specifications of the device are in the commit message. The AC Card is a QCA9880 and the Ethernet PHY is a AR8033.

When i use the 5GHz WiFi of my Archer C58, the device crashes due to OOM. My Archer C7 also uses around 64MB of RAM in operation which would also lead to OOM if the Device would have half the RAM.

blogic commented on target/linux/ar71xx/base-files/etc/board.d/02_network:
```diff
@@ -80,6 +80,7 @@ ar71xx_setup_interfaces()
 	el-mini|\
 	fritz300e|\
+	fritz1750e|\
 	gl-usb150|\
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -712,6 +712,9 @@ ar71xx_board_detect() {
 	*"FRITZ!WLAN Repeater 300E")
 		name="fritz300e"
 		;;
+	*"FRITZ!WLAN Repeater 1750E")
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -685,6 +685,7 @@ platform_check_image() {
 		;;
 	# these boards use metadata images
 	fritz300e|\
+	fritz1750e|\
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/config-4.9:
```diff
@@ -115,6 +115,7 @@ CONFIG_ATH79=y
 # CONFIG_ATH79_MACH_EW_DORIN is not set
 # CONFIG_ATH79_MACH_F9K1115V2 is not set
 # CONFIG_ATH79_MACH_FRITZ300E is not set
+# CONFIG_ATH79_MACH_FRITZ1750E is not set
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt:
```diff
@@ -2347,3 +2347,12 @@ config ATH79_MACH_FRITZ300E
 	select ATH79_DEV_LEDS_GPIO
 	select ATH79_DEV_M25P80
+
+config ATH79_MACH_FRITZ1750E
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt:
```diff
@@ -2347,3 +2347,12 @@ config ATH79_MACH_FRITZ300E
 	select ATH79_DEV_M25P80
+
+config ATH79_MACH_FRITZ1750E
+	bool "AVM FRITZ!WLAN Repeater 1750E support"
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/files/arch/mips/ath79/Makefile:
```diff
@@ -125,6 +125,7 @@ obj-$(CONFIG_ATH79_MACH_EW_BALIN)		+= mach-ew-balin.o
 obj-$(CONFIG_ATH79_MACH_EW_DORIN)		+= mach-ew-dorin.o
 obj-$(CONFIG_ATH79_MACH_F9K1115V2)		+= mach-f9k1115v2.o
 obj-$(CONFIG_ATH79_MACH_FRITZ300E)		+= mach-fritz300e.o
+obj-$(CONFIG_ATH79_MACH_FRITZ1750E)		+= mach-fritz1750e.o
```
alphanumeric ordering please

blogic commented on target/linux/ar71xx/files/arch/mips/ath79/mach-fritz1750e.c:
```diff
@@ -0,0 +1,252 @@
+#define FRITZ1750E_74HC_GPIO_LED_RSSI4		(FRITZ1750E_74HC_GPIO_BASE + 4)
+#define FRITZ1750E_74HC_GPIO_LED_WLAN		  (FRITZ1750E_74HC_GPIO_BASE + 5)
+#define FRITZ1750E_74HC_GPIO_LED_POWER		  (FRITZ1750E_74HC_GPIO_BASE + 6)
+
```
lots of indenting and space/vs tabs errors in this file. please make it all nice and consistent.

blogic commented on target/linux/ar71xx/files/arch/mips/ath79/mach-fritz1750e.c:
```diff
@@ -0,0 +1,252 @@
+	.name = "mdio-gpio",
+	.id = 0,
+	.dev = {
+		.platform_data = &fritz1750e_mdio_data, &fritz1750e_ar8033_data
```
this cant be right.

blocktrron replied:
Honestly, i've just oriented myself how it was handled with the EAP120 ( https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-eap120.c#L84 ). Do you have an idea how to achieve the goal of using the mdio_gpio_platform_data and at803x_platform_data?

blocktrron commented:
@blogic Thanks for your feedback! 

In the meantime I've got the PHY working, the mdc/mdio GPIOs in GPL source were actually pull-ups. Communication with the PHY is fine now. However, if the PHY performas an auto-negotiation resulting in a different speed than the first auto-negotiation after power-up, the kernel detects the link speed correctly but no TX/RX is happening.

This behavior is not happening with a similar device (UBNT AC-Mesh / AT803x/SGMII).

robimarko commented:
That is most likely due to PLL configuration missing

blocktrron commented:
@robimarko I've added PLL configuration from the vendor GPL (drivers/net/avm_cpmac/switch/atheros/atheros_gmac.c L1107) now. However, the PHY is showing the same problem.

Sometimes after linkup, dmesg gets filled with

```
[  165.538503] 803x_aneg_done: SGMII link is not ok
[  166.578501] 803x_aneg_done: SGMII link is not ok
[  167.618502] 803x_aneg_done: SGMII link is not ok
[  168.658503] 803x_aneg_done: SGMII link is not ok
[  169.698501] 803x_aneg_done: SGMII link is not ok
[  170.738501] 803x_aneg_done: SGMII link is not ok
[  171.778502] 803x_aneg_done: SGMII link is not ok
[  172.818502] 803x_aneg_done: SGMII link is not ok
[  173.858501] 803x_aneg_done: SGMII link is not ok
[  174.898502] 803x_aneg_done: SGMII link is not ok
[  175.938501] 803x_aneg_done: SGMII link is not ok
[  176.978502] 803x_aneg_done: SGMII link is not ok
[  178.018503] 803x_aneg_done: SGMII link is not ok
[  179.058516] 803x_aneg_done: SGMII link is not ok
[  180.098501] 803x_aneg_done: SGMII link is not ok
[  181.138498] 803x_aneg_done: SGMII link is not ok
[  182.178502] 803x_aneg_done: SGMII link is not ok
[  183.218496] 803x_aneg_done: SGMII link is not ok
[  184.258502] 803x_aneg_done: SGMII link is not ok
[  185.298501] 803x_aneg_done: SGMII link is not ok
[  186.338501] 803x_aneg_done: SGMII link is not ok
[  187.378513] 803x_aneg_done: SGMII link is not ok
[  188.418501] 803x_aneg_done: SGMII link is not ok
[  189.458501] 803x_aneg_done: SGMII link is not ok
```

TX/RX is not working in this case. This error only occurs if link speed has changed. However, i'm not able to reproduce it reliably.

Could the PLL configuration still be incorrect or might this indicate another problem? This error occurs according to the kernel source (if im not mistaken) in case the PHY doesn't pull the SGMII link up.

robimarko commented:
Have you tried adding ath79_eth0_data.duplex = DUPLEX_FULL;

blocktrron commented:
Not yet, will try.

blocktrron commented:
@robimarko thanks for you patience. Iäve tried adding `ath79_eth0_data.duplex = DUPLEX_FULL;` today. Unfortunately, the bahivour has not changed iin any way.

blocktrron commented:
@robimarko Thanks for your hint regarding the PLL settings. In the end, they were indeed part of the problem but furthermore they were written into the wrong register (see my first commit). 10/100/1000 Ethernet is working now.

@blogic Is anything else left regarding merging this? The 5GHz radio is not planned (from my side atleast).

robimarko commented:
I dont think that it should be merged without 5GHz radio working as ath10k driver supports it and is used on devices with similar specs

blocktrron commented:
I've looked into the 5GHz module some time ago, but the thing is i'm not able to get the kernel to recognize it. An image with pciutils built-in returns in no output on `lspci`. 

If you want to give it a shot, my attempt is located here, maybe something very obvious is missing: https://github.com/blocktrron/openwrt/commit/429579019eb560f9962b82d67b89548efcfc6bfa . 

I'm aware anything calibration-data related is missing, which i've didn't even bother due to the module not showing up in the first place. 

I'm repeating my statement from some comments ago, that i've experienced ath10k in combination with 64MB as very unstable in noisy environments (Archer C25, Archer C58, Archer C60). Therefore, the integration of the 5GHz module is not a top priority for me.

ecsv commented:
blocktrron asked me to add some infos here (previously mentioned in the #gluon IRC channel):

 64MB is really, really low for ath9k+ath10k but it could still work when using https://github.com/openwrt/openwrt/blob/751746c736bbda828ff1add002a0419a3b883e63/package/kernel/mac80211/patches/960-0010-ath10k-limit-htt-rx-ring-size.patch and 
https://github.com/openwrt/openwrt/blob/751746c736bbda828ff1add002a0419a3b883e63/package/kernel/mac80211/patches/960-0011-ath10k-limit-pci-buffer-size.patch and reducing the queue-lengths (including the aqm limits as shown in https://github.com/freifunk-gluon/gluon/commit/e63c6ca01f50c96d76e5570faa290617a8a312b4 ).

But this has to be verified on such a device. I can only say for sure that a 256MB device with 3x ath10k rather often gets an OOM when stressing the device (for example by transferring a lot of multicast traffic and multiple unicast streams). The previously mentioned changes helped to reduce this problem.

blocktrron committed:
ar71xx: fix QCA955X SGMII link loss

The QCA955X is affected by a hardware bug which causes link-loss of the
SGMII link between SoC and PHY. This happens on change of link-state or
speed.

It is not really known what causes this bug. It definitely occurs when
using a AR8033 Gigabit Ethernet PHY.

This bug was first proposed to be fixed by Sven Eckelmann in 2016.
 https://patchwork.ozlabs.org/patch/604782/

Based-on-patch-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>
Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
ar71xx: allow to override at803x sgmii aneg status

When checking the outcome of the PHY autonegotiation status, at803x
currently returns false in case the SGMII side is not established.

Due to a hardware-bug, ag71xx needs to fixup the SoCs SGMII side, which
it can't as it is not aware of the link-establishment.

This commit allows to ignore the SGMII side autonegotiation status to
allow ag71xx to do the fixup work.

Signed-off-by: David Bauer <mail@david-bauer.net>

ynezz commented:
Closing as we've agreed on IRC. Thanks!
